### ISSUE TITLE
Is getting version from environment var legit?

### ISSUE THREAD

**Issue by sourcedelica** [conan-io/conan#565]
Currently I'm getting the version using `os.environ`.  This is to avoid needing to edit the Conanfile for each new version of the library.  Also it's to ensure that the `test_package` Conanfile is using the same value as its parent requirement.   Otherwise we'd need to edit _two_ Conanfiles for each new version, a recipe (hah!) for disaster.

I wanted to make sure this is legit, that it wouldn't cause problems.

**Comment by memsharded** [comment]
I am afraid it will not work for recipes, when they are loaded in another user machine, and the environment variable is not set. Furthermore, it would be problematic and error prone if managing more than one version of the same library in the same machine. Note that private dependencies are allowed, for example for building tests, one library can depend on gtest/1.7, and another one on gtest/1.8. If both are based on the same recipe and use an environment variable, even if defining the environment variable the build would be broken.

My suggestion: the package manager is an intermediate piece of the automation stack. Below is the build system. Above is the build/project/CI manager... Maybe it should manage it. The simplest approach would be a template with a placeholder to search&replace to generate the real files. 

Latest suggestion from @Prinkesh in https://github.com/conan-io/conan/issues/548#issuecomment-253974352 of a ``exports()`` method to automate tasks might help too. We will think specifically about this use case of the version.

**Comment by sourcedelica** [comment]
Ok.  Thanks.   The layer on top can roll the version number.  That's how Maven's release plugin works.

**Comment by sourcedelica** [comment]
Thanks to the pointer to #548. Definitely relevant.  I have been meaning to try `exports =`.  I wasn't sure if it was mutually exclusive from the `Packaging from external sources` approach.  But it looks like it's not.  I added a `version.txt` file and exported it and it seemed to work for exporting the package and building it on `install`.  A `version.txt` would be easier for a program to update than the Conanfile.

Should you be able to use `requirements()` in `test_package/conanfile.py`?  I kept on getting the error `ERROR: Unable to retrieve first requirement of test conanfile.py`.  For example:

    def requirements(self):
        self.requires("caf/0.15.1@sourcedelica/testing")
